enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,199 @@
+---
+title: loadbalancer-service-support
+authors:
+  - "@pliurh"
+reviewers:
+  - "@copejon, MicroShift contributor"
+  - "@fzdarsky, MicroShift architect"
+  - "@ggiguash, MicroShift contributor"
+  - "@dhellmann, MicroShift contributor"
+  - "@oglok, MicroShift contributor"
+  - "@zshi-redhat, MicroShift "
+approvers:
+  - "@dhellmann"
+api-approvers:
+  - None
+creation-date: 2023-01-13
+last-updated: 2023-01-13
+tracking-link:
+  - https://issues.redhat.com/browse/NP-604
+---
+
+# MicroShift Service of Loadbalancer Type Support
+
+## Summary
+
+Customers using loadbalancer type Service, doesn't want to make changes to their
+workloads/deployment for MicroShift as they use other k8s flavors as well. It is
+also promised by us that users do not have to modify workload manifests that
+work on OpenShift to run on MicroShift (as long as they do not use APIs
+unsupported by MicroShift).
+
+## Motivation
+
+Normally, the loadbalancer type Service is supported by a Kubernetes cluster
+with external load balancers. In Microshift, as there is only one node in the
+cluster, we don't really need to have a load balancer. However, from user
+experience perspective, it's important for customers to be able to migrate
+the workload to Microshift without modification. So we need to find a way to
+support the loadbalancer type Service without a load balancer.
+
+### User Stories
+
+As a Microshift user, I want to reuse the Service manifest that with type of
+`LoadBalancer` without modification. I don't want to allocate a new IP address
+to the Node as the loadbalancer VIP. I want to use the NodeIP as the ingress IP
+of the loadbalancer Service. 
+
+### Goals
+
+* Allow users to deploy the load balancer type Service to Microshift.
+* The node IP shall be used as the ingress IP of the Service.
+* As we will support pluggable CNI in Microshift. The implementation shall be
+also compatible with CNIs other than OVN-Kubernetes.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+To support the loadbalancer type Service, we need to create a new controller
+component that watch the Service objects. It shall be responsible for:
+
+1. work with CNIs to plumb the ingress traffic to the endpoint pods.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+### Workflow Description
+
+**cluster user** is a human user responsible for provisioning Services to a
+Microshift cluster.
+
+1. The cluster user deploys a Service of type: LoadBalancer 
+2. The loadbalancer service controller watches the creation of the service. It
+will update the Service's `.status.loadBalancer` field, and help the CNIs to 
+plumbing the ingress traffic to the endpoint pods.
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+This new controller is part of the microshift binary. It shall be able to know
+what CNI is used by the cluster, and behave differently based on the CNI type.
```

Comment:
I am not sure if MicroShift may have this type of visibility into the which plugin, we certainly could, but we would be adding again coupling (microshift to CNI). See my above comments, I wonder if they make sense.

Do you have a link to the code on OVNk handling this plumbing?

Reply:
https://github.com/openshift/ovn-kubernetes/blob/master/docs/external-ip-and-loadbalancer-ingress.md

https://github.com/ovn-org/ovn-kubernetes/blob/bbaa7d702103a912dcfe75d3b6292bfd18ecaf74/go-controller/pkg/node/gateway_shared_intf.go#L240

Reply:
It is hard to say. It depends on CNI datapath implementation. For those CNIs that don't either bypass the kernel or have their own implementation like OVN-K, I think we can do plumbing by adding iptables rules. I don't think there is a universal implementation that can work with any CNI.

Reply:
K3s uses klipper-lb ([code](https://github.com/k3s-io/klipper-lb/blob/master/entry)) in form of DaemonSet for each service. Log for `8080`:
```
+ trap exit TERM INT
+ echo 0.0.0.0/0
+ grep -Eq :
+ iptables -t filter -I FORWARD -s 0.0.0.0/0 -p TCP --dport 8080 -j ACCEPT
+ echo 10.43.68.10
+ grep -Eq :
+ cat /proc/sys/net/ipv4/ip_forward
+ '[' 1 '==' 1 ]
+ iptables -t filter -A FORWARD -d 10.43.68.10/32 -p TCP --dport 8080 -j DROP
+ iptables -t nat -I PREROUTING '!' -s 10.43.68.10/32 -p TCP --dport 8080 -j DNAT --to 10.43.68.10:8080
+ iptables -t nat -I POSTROUTING -d 10.43.68.10/32 -p TCP -j MASQUERADE
+ '[' '!' -e /pause ]
+ mkfifo /pause
```
`10.43.68.10` is `clusterIP` so it's just forwarding ports from host to clusterIP. There's a chance that something similar should work for any CNI, right? Maybe later we can consider disabling ovn-k handling and have the same logic for every CNI.

Reply:
What if a CNI has already got LB service support similar to OVN-K. Do we also modify the CNI to turn it off? And for some CNI implementations, their datapath will bypass the kernel stack, or use hardware offload like DPU. How can we use iptables rules to do the job?

Reply:
@pmtk , @pliurh , can we resolve this thread?